Add TLS support to Torii

ionagamed opened the pull request:
Signed-off-by: Leonid Lygin <ionagamed@gmail.com>

<!-- You will not see HTML commented line in Pull Request body -->
<!-- Optional sections may be omitted. Just remove them or write None -->

<!-- ### Requirements -->
<!-- * Filling out the template is required. Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion. -->
<!-- * All new code must have code coverage above 70% (https://docs.codecov.io/docs/about-code-coverage). -->
<!-- * CircleCI builds must be passed. -->
<!-- * Critical and blocker issues reported by Sorabot must be fixed. -->
<!-- * Branch must be rebased onto base branch (https://soramitsu.atlassian.net/wiki/spaces/IS/pages/11173889/Rebase+and+merge+guide). -->


### Description of the Change

This adds optional TLS encryption for torii on a separate port

<!-- We must be able to understand the design of your change from this description. If we can't get a good idea of what the code will be doing from the description here, the pull request may be closed at the maintainers' discretion. -->
<!-- Keep in mind that the maintainer reviewing this PR may not be familiar with or have worked with the code here recently, so please walk us through the concepts. -->

### Benefits

Improved security of client-node communication

<!-- What benefits will be realized by the code change? -->

### Possible Drawbacks 

None

<!-- What are the possible side-effects or negative impacts of the code change? -->
<!-- If no drawbacks, explicitly mention this (write None) -->

### Usage Examples or Tests *[optional]*

<!-- Point reviewers to the test, code example or documentation which shows usage example of this feature -->

### Alternate Designs *[optional]*

<!-- Explain what other alternates were considered and why the proposed version was selected -->

igor-egorov commented on docs/source/guides/sec-install.rst:
```diff
@@ -21,12 +21,15 @@ Create an encrypted backup of private keys before deleting them and limit the ac
 Ideally, Iroha should be as much isolated as possible in terms of networking.
 
-Currently, there is no traffic encryption in Iroha, we strongly recommend using VPN or Calico for setting up Docker overlay network, i.e. any mechanism that allows encrypting communication between peers.
+You can enable TLS on torii ports if you would like to encrypt client-peer 
```
```suggestion
You can enable TLS on torii ports if you would like to encrypt client-peer
```
(trailing whitespace)

igor-egorov commented on docs/source/guides/torii-tls.rst:
```diff
@@ -0,0 +1,51 @@
+
+If you need to use plain IP addresses to connect to the node, you need to
+specify subjectAltName in your server certificate, for that you need to add
+a ``subjectAltName`` directive to ``v3_ca`` section of your openssl config 
```
Please install some plugin to your editor that shows or automatically removes trailing whitespaces.
```suggestion
You can enable TLS on torii ports if you would like to encrypt client-peer
```

ionagamed replied:
Fixed in latest version

igor-egorov commented on docs/source/guides/torii-tls.rst:
```diff
@@ -0,0 +1,51 @@
+If you need to use plain IP addresses to connect to the node, you need to
+specify subjectAltName in your server certificate, for that you need to add
+a ``subjectAltName`` directive to ``v3_ca`` section of your openssl config 
+before generating the certificate. 
```
trailing whitespace
```suggestion
before generating the certificate.
```

ionagamed replied:
Fixed in latest version

igor-egorov commented on docs/source/guides/torii-tls.rst:
```diff
@@ -0,0 +1,51 @@
+
+Fields in the certificate don't really matter except for the Common Name (CN),
+it would be checked against the client's hostname, and TLS handshake will fail
+if they do not match (e.g. if you connect to example.com:50051, then irohad at 
```
```suggestion
if they do not match (e.g. if you connect to example.com:50051, then irohad at
```
trailing whitespace

ionagamed replied:
Fixed in latest version

igor-egorov commented on docs/source/guides/torii-tls.rst:
```diff
@@ -0,0 +1,51 @@
+
+``torii_tls_keypair`` - set this to full path to the key/certificate pair,
+so if you have a key at ``/path/to/server.key`` and a certificate at
+``/path/to/server.crt``, you need to specify 
```
```suggestion
``/path/to/server.crt``, you need to specify
```
trailing whitespace

ionagamed replied:
Fixed in latest version

igor-egorov commented on irohad/main/application.hpp:
```diff
@@ -80,6 +80,8 @@ class Irohad {
    * @param listen_ip - ip address for opening ports (internal & torii)
    * @param torii_port - port for torii binding
+   * @param torii_tls_port - port for torii TLS binding
+   * @param torii_tls_keypair - path to TLS keypair (.crt, .key) for torii
```
Please add a remark about the path format. Should it include file extension or not?

ionagamed replied:
Fixed in latest version

igor-egorov commented on irohad/main/application.hpp:
```diff
@@ -80,6 +80,8 @@ class Irohad {
    * @param listen_ip - ip address for opening ports (internal & torii)
    * @param torii_port - port for torii binding
+   * @param torii_tls_port - port for torii TLS binding
+   * @param torii_tls_keypair - path to TLS keypair (.crt, .key) for torii
```
Please add a remark about the path format. Should it include file extension or not?

ionagamed replied:
Fixed in latest version

igor-egorov commented on irohad/main/application.cpp:
```diff
@@ -741,6 +745,15 @@ Irohad::RunResult Irohad::run() {
+  if (enable_tls) {
+    torii_tls_server = std::make_unique<ServerRunner>(
+        listen_ip_ + ":" + std::to_string(torii_tls_port_),
+        log_manager_->getChild("ToriiServerRunner")->getLogger(),
```
```suggestion
        log_manager_->getChild("ToriiTlsServerRunner")->getLogger(),
```

ionagamed replied:
Fixed in latest version

igor-egorov commented on irohad/main/server_runner.hpp:
```diff
@@ -24,7 +24,8 @@ class ServerRunner {
                         logger::LoggerPtr log,
-                        bool reuse = true);
+                        bool reuse = true,
+                        const std::string &tls_keypair = "");
```
``tls_keypair`` argument is not documented in the docstring above. Please add a comment.

ionagamed replied:
Fixed in latest version

igor-egorov commented on irohad/main/server_runner.hpp:
```diff
@@ -64,6 +65,7 @@ class ServerRunner {
   std::string serverAddress_;
   bool reuse_;
   std::vector<std::shared_ptr<grpc::Service>> services_;
+  std::string tlsKeypair_;
```
We use ``camel_case_`` naming style for internal class members.
```suggestion
  std::string tls_keypair_;
```

ionagamed replied:
There is `serverAddress_` literally 4 lines before this one, so I was confused :)

ionagamed replied:
Fixed in latest version

igor-egorov commented on irohad/main/server_runner.cpp:
```diff
@@ -5,15 +5,22 @@
 
 #include "main/server_runner.hpp"
 
+#include <fstream>
```
According to the Google c++ style guide any STL includes have to be divided from the others with a blank line.
Please add a blank line.

https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes

ionagamed replied:
Fixed

igor-egorov commented on irohad/main/server_runner.cpp:
```diff
@@ -5,15 +5,22 @@
+    : log_(std::move(log)),
+      serverAddress_(address),
+      reuse_(reuse),
+      tlsKeypair_(tls_keypair) {}
```
tls_keypair_. serverAddress_ can be refactored too since you have been touched that part of the code :)

ionagamed replied:
Fixed in latest version

igor-egorov commented on irohad/main/server_runner.cpp:
```diff
@@ -5,15 +5,22 @@
+    : log_(std::move(log)),
+      serverAddress_(address),
+      reuse_(reuse),
+      tlsKeypair_(tls_keypair) {}
```
tls_keypair_. serverAddress_ can be refactored too since you have been touched that part of the code :)

ionagamed replied:
Fixed in latest version

igor-egorov commented on irohad/main/server_runner.cpp:
```diff
@@ -28,8 +35,37 @@ iroha::expected::Result<int, std::string> ServerRunner::run() {
+    options.pem_key_cert_pairs.push_back(keypair);
+    auto credentials = grpc::SslServerCredentials(options);
+    builder.AddListeningPort(serverAddress_, credentials, &selected_port);
+  } else { // if (enable_tls)
```
```suggestion
  } else { // tls is not enabled
```

ionagamed replied:
Fixed in latest version

igor-egorov commented on irohad/main/server_runner.cpp:
```diff
@@ -28,8 +35,37 @@ iroha::expected::Result<int, std::string> ServerRunner::run() {
-  builder.AddListeningPort(
-      serverAddress_, grpc::InsecureServerCredentials(), &selected_port);
+  bool enable_tls = tlsKeypair_.length() > 0;
+  if (enable_tls) {
```
Probably the whole TLS initialization code can be moved to a separate method.
Please consider that option. Otherwise, ``run`` method comes quite big.

ionagamed replied:
Fixed in latest version

igor-egorov commented on scripts/run-iroha-dev.sh:
```diff
@@ -18,16 +18,19 @@ export COMPOSE_PROJECT_NAME=${PROJECT}
 else
-    IROHA_DBG_PORTS="$(docker port ${PROJECT}_node_1 | sed 's/\(.*\)://' | sort -r | sed -e :a -e N -e 's/\n/:/p' -e ta)"
-    export IROHA_PORT="$(echo ${IROHA_DBG_PORTS} | sed 's/:.*//')"
+    IROHA_DBG_PORTS="$(docker port ${PROJECT}_node_1 | sed 's/\(.*\)://' | sort -r | paste -s -d ':' -)"
```
👀

igor-egorov commented on test/framework/integration_framework/integration_test_framework.cpp:
```diff
@@ -154,6 +156,8 @@ namespace integration_framework {
                                             block_store_path,
                                             kLocalHost,
                                             torii_port_,
+                                            0,
```
Why you are not using here the value of previously defined constant ``constexpr size_t kDefaultToriiTlsPort = 0;``?
I believe that will make the code more clear.

ionagamed replied:
Fixed in latest version

igor-egorov commented on test/framework/integration_framework/integration_test_framework.cpp:
```diff
@@ -154,6 +156,8 @@ namespace integration_framework {
                                             kLocalHost,
                                             torii_port_,
+                                            0,
+                                            "",
```
Not cool.
Probably we need some struct like this:
```
struct TlsConfig {
    size_t port,
    std::string key_path
};
```
Then, two initialization parameters can be substituted by the structure.
Moreover, looks like that combined argument can be declared as ``boost::optional`` because it is not a required initialization argument.

ionagamed replied:
Fixed in latest version. This struct is available as `iroha::torii::TlsParams`

igor-egorov commented on test/framework/integration_framework/iroha_instance.hpp:
```diff
@@ -47,6 +47,8 @@ namespace integration_framework {
                   const std::string &block_store_path,
                   const std::string &listen_ip,
                   size_t torii_port,
+                  size_t torii_tls_port,
```
Look at my comment in `integration_test_framework.cpp` about possible `struct TlsConfig`;

igor-egorov replied:
Newly introduced arguments are not documented in a docstring above.

ionagamed replied:
Fixed in latest version

ionagamed commented on docs/source/guides/torii-tls.rst:
```diff
@@ -0,0 +1,51 @@
+By default, client-peer communication is not encrypted.
+To enable it, you need to:
+
+1. `Generate <Generating keys>`_ a key/certificate pair for each peer
```
Fixed in latest version, they did not compile correctly, but now do

ionagamed commented on docs/source/guides/torii-tls.rst:
````diff
@@ -0,0 +1,51 @@
+Keys must be presented in PEM format. To generate them you can use ``openssl``:
+
+```bash
+$ openssl genpkey -algorithm <desired algorithm> -out server.key
````
Fixed in latest version

ionagamed commented on docs/source/guides/torii-tls.rst:
````diff
@@ -0,0 +1,51 @@
+```conf
+[ v3_ca ]
+subjectAltName=IP:12.34.56.78
+...
````
Fixed in latest version

ionagamed commented on docker/docker-compose.yml:
```diff
@@ -5,6 +5,7 @@ services:
     image: hyperledger/iroha:develop-build
     ports:
       - "${IROHA_PORT}:50051"
+      - "${IROHA_TLS_PORT}:50052"
```
Fixed in latest version

ionagamed replied:
Fixed in latest version. Port was reassigned to 55552

ionagamed commented:
Fixed all the comments. This is ready for a re-review

MBoldyrev commented on irohad/main/iroha_conf_literals.hpp:
```diff
@@ -14,6 +14,8 @@
   extern const char *BlockStorePath;
   extern const char *ToriiPort;
+  extern const char *ToriiTlsPort;
+  extern const char *ToriiTlsKeypair;
```
add
```c++
extern const char *ToriiTlsParams
```

MBoldyrev commented on irohad/main/iroha_conf_loader.cpp:
```diff
@@ -355,6 +355,9 @@ inline void JsonDeserializerImpl::getVal<IrohadConfig>(
   const auto obj = src.GetObject();
   getValByKey(path, dest.block_store_path, obj, config_members::BlockStorePath);
   getValByKey(path, dest.torii_port, obj, config_members::ToriiPort);
+  getValByKey(path, dest.torii_tls_port, obj, config_members::ToriiTlsPort);
```
```suggestion
  getValByKey(path, dest.torii_tls_params, obj, config_members::ToriiTlsParams);
```

ionagamed commented:
Fixed all previous comments from @MBoldyrev, now the config can have

```json
{
    "torii_tls_params": {
        "port": 55552,
        "key_pair_path": "/path/to/a/keypair"
    }
}
```

igor-egorov commented:
need to update branch

ionagamed commented:
Rebased on top of target branch

lebdron commented on irohad/main/server_runner.hpp:
```diff
@@ -44,6 +48,23 @@ class ServerRunner {
+   * @return Result with server credentials or error message
+   */
+  iroha::expected::Result<std::shared_ptr<grpc::ServerCredentials>, std::string>
+  createSecureCredentials();
```
There methods should definitely not be public. Please make them `private` or `protected`, or move them to an anonymous namespace in cpp file.

lebdron commented on test/system/irohad_test.cpp:
```diff
@@ -157,6 +158,42 @@ class IrohadTest : public AcceptanceFixture {
+          kAddress + ":" + std::to_string(port),
+          (test_data_path_ / "tls/correct.crt").string())
+          .match([&](auto secure_stub) { stub = std::move(secure_stub.value); },
+                 [&](const auto &error) { stub_error = error.error; });
```
Maybe print `stub_error` somewhere or return it?

MBoldyrev commented on irohad/main/server_runner.hpp:
```diff
@@ -57,13 +78,14 @@ class ServerRunner {
-  std::condition_variable serverInstanceCV_;
+  std::unique_ptr<grpc::Server> server_instance_;
+  std::mutex wait_for_server_;
+  std::condition_variable server_instance_cv_;
```
:snake: :+1:

lebdron commented on test/system/irohad_test.cpp:
```diff
@@ -157,6 +158,42 @@ class IrohadTest : public AcceptanceFixture {
+    }
+
+    return torii::CommandSyncClient(
+        std::move(stub),
```
It does not look like it will work if there was an error in `createSecureClient`.

lebdron commented on test/system/irohad_test.cpp:
```diff
@@ -297,6 +333,48 @@ TEST_F(IrohadTest, SendTx) {
+ * @given running Iroha with an open TLS port
+ * @when a client sends a transaction to Iroha AND the server's certificate
+ *       is valid
+ * @then the transaction is commited
```
```suggestion
 * @then the transaction is committed
```
same below

MBoldyrev commented on test/system/irohad_test.cpp:
```diff
@@ -157,6 +158,42 @@ class IrohadTest : public AcceptanceFixture {
+        port = kSecurePort;
+      } else {
+        port = kPort;
+      }
```
```c++
    uint16_t port = override_port.value_or(enable_tls ? kSecurePort : kPort);
```
instead of 9 lines above

MBoldyrev commented on test/system/irohad_test_data/config.sample:
```diff
@@ -1,8 +1,12 @@
+  },
   "internal_port" : 10001,
-  "pg_opt" : "host=localhost port=5432 user=postgres password=mysecretpassword",
+  "pg_opt" : "host=localhost port=15432 user=iroha password=helloworld",
```
why is this change needed?

lebdron commented:
Please also check `stop-iroha-dev.sh`

ionagamed commented:
Fixed all the comments, except the one about adding another port to the `ServerRunner`.

Now the arguments are passed through the constructor, and then everything happens in the `run()` method. There are two ways I see that multiple ports could be implemented in `ServerRunner`:
1. Pass it through the constructor, via an optional container for both port and 
    certificate (e.g. `iroha::torii::TlsParams`).
2. Expose the interface to add as many listening ports, e.g.
    ```cpp
    iroha::expected::Result<...> addListeningPort(uint16_t, boost::optional<TLSKeypair>)
    ```
And I'm not sure which one to implement.

Otherwise, this is ready for a re-review.

MBoldyrev commented on example/config.docker:
```diff
@@ -2,7 +2,7 @@
   "torii_port" : 50051,
   "internal_port" : 10001,
-  "pg_opt" : "host=some-postgres port=5432 user=postgres password=mysecretpassword",
+  "pg_opt" : host=some-postgres port=5432 user=postgres password=mysecretpassword",
```
whoops

MBoldyrev commented on irohad/main/CMakeLists.txt:
```diff
@@ -5,7 +5,7 @@
 set(EXECUTABLE_OUTPUT_PATH ${PROJECT_BINARY_DIR}/bin)
 
-add_library(server_runner server_runner.cpp)
+add_library(server_runner server_runner.cpp tls_keypair.cpp)
```
please follow the multiline style for targets with multiple source files

MBoldyrev commented on irohad/main/application.hpp:
```diff
@@ -14,9 +14,11 @@
 #include "interfaces/queries/query.hpp"
 #include "logger/logger_fwd.hpp"
 #include "logger/logger_manager_fwd.hpp"
+#include "main/server_runner.hpp"
```
why did you change the forward declaration for header inclusion?

ionagamed replied:
`boost::optional<std::shared_ptr<ServerRunner>>` seems to require a full type, otherwise my compiler gives `error: invalid application of 'sizeof' to an incomplete type 'ServerRunner'` (clang-10.0.1 on macos)

MBoldyrev replied:
Tested it. Seems you are right, and I am not. Then this include is OK.

MBoldyrev commented on irohad/main/server_runner.cpp:
```diff
@@ -39,36 +47,68 @@ iroha::expected::Result<int, std::string> ServerRunner::run() {
+}
+
+iroha::expected::Result<std::shared_ptr<grpc::ServerCredentials>, std::string>
+ServerRunner::createSecureCredentials() {
```
is this function supposed to ever return *error*? In its body i can see no such option. If so, why wrapping returned object into `Result`?

MBoldyrev commented on irohad/main/server_runner.cpp:
```diff
@@ -39,36 +47,68 @@ iroha::expected::Result<int, std::string> ServerRunner::run() {
+      return credentials_error;
+    }
+
+    builder.AddListeningPort(server_address_, credentials, selected_port);
```
How are the keys having incorrect format handled? I am afraid, an exception may arise on this line or somewhere later. If this is the case, should catch such an exception and provide the user with an appropriate error message.
Also please check what happens if we provide a syntactically correct key that does not match the certificate.

ionagamed replied:
I could not yet find how to handle this error getting a cause from gRPC, but invalid formats and mismatched certificates would both give `selected_port == 0`, and would print an error message, such as `E0714 18:22:43.053279000 4790195648 ssl_transport_security.cc:675]     Invalid cert chain file.` What would you suggest, do I keep digging (I have literally spent the weekend looking for a solution), or this would be sufficient?

MBoldyrev replied:
I think that if you checked various error cases and irohad 1) points up the correct error reason and 2) terminates gracefully - that is enough for now. Thank you for devoting your time to the investigation!

MBoldyrev commented on test/framework/integration_framework/iroha_instance.hpp:
```diff
@@ -17,6 +17,7 @@
 #include "logger/logger_fwd.hpp"
 #include "logger/logger_manager_fwd.hpp"
 #include "multi_sig_transactions/gossip_propagation_strategy_params.hpp"
+#include "torii/tls_params.hpp"
```
Types wrapped into boost::optional suffice with a forward declaration.

MBoldyrev replied:
Same as above, my fault.

MBoldyrev commented:
Thanks!

> I'm not sure which one to implement.

My vote is for the second, although it could be done in a separate PR as a refactoring task.

ionagamed commented:
Fixed all previous comments (missing quote, redundant `Result` and `CMakeLists.txt` style). Probably ready for a re-review (I hope :) )

lebdron commented on irohad/main/tls_keypair.cpp:
```diff
@@ -0,0 +1,22 @@
+}
+
+TLSKeypair::TLSKeypair(const std::string &path)
+    : certificate(readFile(path + ".crt")),
```
Basically these two `readFile` calls can be done somewhere in irohad.cpp, then two `Blob` objects are initialized from `readFile` result, and `TLSKeypair` is initialized from two `Blob` objects. This will allow to initialize `ServerRunner` not only from files.

ionagamed replied:
Oh, okay, nevermind the last comment, this seems more reasonable, will do so

ionagamed commented on irohad/main/server_runner.hpp:
```diff
@@ -21,10 +22,13 @@ class ServerRunner {
+      const std::string &address,
+      logger::LoggerPtr log,
+      bool reuse = true,
+      const boost::optional<TLSKeypair> &tls_keypair = boost::none);
```
`TLSKeypair` takes on the responsibility of reading the files, so I can maybe add a separate constructor to create a keypair from raw data?

ionagamed commented:
Fixed the docs (it is now a valid json structure), and moved the file-reading logic into a separate `TLSKeypairFactory`. I didn't see much use in going with a `Blob` instead of `std::string`, because PEM-encoded keys are printable-text-only anyway, and gRPC requires the keys in PEM form

lebdron commented on example/torii_tls/server.crt:
```diff
@@ -0,0 +1,16 @@
+-----BEGIN CERTIFICATE-----
```
Bump

lebdron commented on irohad/main/tls_keypair.hpp:
```diff
@@ -8,17 +8,32 @@
 
-#endif // IROHA_TLS_KEYPAIR_HPP
\ No newline at end of file
+#endif  // IROHA_TLS_KEYPAIR_HPP
```
A newline would be great.
